pkg/util/kubernetes/kubelet/kubelet.go

```diff
@@ -125,58 +131,130 @@ func (ku *KubeUtil) searchPodForContainerID(podlist []*Pod, containerID string)
 	return nil, fmt.Errorf("container %s not found in podlist", containerID)
 }
 
-// Try and find the hostname to query the kubelet
-// TODO: Add TLS verification
-func locateKubelet() (string, error) {
-	host := config.Datadog.GetString("kubernetes_kubelet_host")
-	if host == "" {
-		var err error
-		host, err = docker.HostnameProvider("")
-		if err != nil {
-			return "", fmt.Errorf("unable to get hostname from docker, please set the kubernetes_kubelet_host option: %s", err)
-		}
+// setupKubeletApiClient will try to setup the http(s) client to query the kubelet
+// with the following settings, in order:
+//  - Load Certificate Authority if needed
+//  - HTTPS w/ configured certificates
+//  - HTTPS w/ configured token
+//  - HTTPS w/ service account token
+//  - HTTP
+func (ku *KubeUtil) setupKubeletApiClient() error {
+	tlsConfig, err := getTLSConfig()
+	if err != nil {
+		return err
 	}
+	transport := &http.Transport{
+		TLSClientConfig: tlsConfig,
+	}
+	ku.kubeletApiClient.Transport = transport
+
+	switch {
+	case isConfiguredCertificates():
+		tlsConfig.Certificates, err = kubernetes.GetCertificates(
+			config.Datadog.GetString("kubelet_client_crt"),
+			config.Datadog.GetString("kubelet_client_key"),
+		)
+		return err
+
+	case isConfiguredTokenPath():
+		return ku.setBearerToken(config.Datadog.GetString("kubelet_auth_token_path"))
 
-	port := config.Datadog.GetInt("kubernetes_http_kubelet_port")
-	url := fmt.Sprintf("http://%s:%d", host, port)
-	healthzURL := fmt.Sprintf("%s%s", url, KubeletHealthPath)
-	if _, err := PerformKubeletQuery(healthzURL); err == nil {
-		return url, nil
+	case kubernetes.IsServiceAccountToken():
+		return ku.setBearerToken(kubernetes.ServiceAccountTokenPath)
+
+		// Without Token and without certificates
+	default:
+		return nil
 	}
-	log.Debugf("Couldn't query kubelet over HTTP, assuming it's not in no_auth mode.")
+}
 
-	port = config.Datadog.GetInt("kubernetes_https_kubelet_port")
-	url = fmt.Sprintf("https://%s:%d", host, port)
-	healthzURL = fmt.Sprintf("%s%s", url, KubeletHealthPath)
-	if _, err := PerformKubeletQuery(healthzURL); err == nil {
-		return url, nil
+func (ku *KubeUtil) setBearerToken(tokenPath string) error {
+	token, err := kubernetes.GetBearerToken(tokenPath)
+	if err != nil {
+		return err
 	}
+	ku.kubeletApiRequestHeaders.Set("Authorization", token)
+	return nil
+}
+
+func (ku *KubeUtil) createKubeletRequest(path string) (*http.Request, error) {
```

Comment:
can that be inlined too, unless the method is to be reused or unit tested?

Most of the codebase keeps the flow in a single method to ease reading, and separate "sections" with a comment line per section. This makes reading a module's logic pretty fast, see https://github.com/DataDog/datadog-agent/blob/master/pkg/collector/autodiscovery/templatecache.go#L78-L107

Reply:
The abstraction is probably too thin but I'm disagree about generally saying:
> be inlined too, unless the method is to be reused or unit tested?

Sometimes you want to abstract a behavior because it's more convenient.